ast.go

```diff
@@ -1977,3 +2065,85 @@ func (node *Revoke) walkSubtree(visit Visit) error {
 	}
 	return Walk(visit, node.Privileges, node.Table)
 }
+
+// resolvers
+
+func resolveReadStatementWalk(node Node, resolver ReadStatementResolver) (string, error) {
+	err := Walk(func(node Node) (bool, error) {
+		if funcExpr, ok := node.(*CustomFuncExpr); ok && funcExpr != nil {
+			resolvedString, err := resolveReadStatement(funcExpr, resolver)
+			if err != nil {
+				return true, err
+			}
+			funcExpr.ResolvedString = resolvedString
+		}
+		return false, nil
+	}, node)
+	if err != nil {
+		return "", err
+	}
+	return node.String(), nil
+}
+
+func resolveReadStatement(node *CustomFuncExpr, resolver ReadStatementResolver) (string, error) {
+	switch node.Name {
+	case "block_num":
+		if len(node.Args) != 1 {
+			return "", errors.New("block_num function should have exactly one argument")
+		}
+		value, ok := node.Args[0].(*Value)
+		if !ok {
+			return "", errors.New("argument of block_num is not a literal value")
+		}
+
+		chainID, err := strconv.ParseInt(string(value.Value), 10, 64)
+		if err != nil {
+			return "", err
```

Comment:
this should never be reached as well. because the argument should always be an integer

Reply:
So this is because in the grammar we have `| BLOCK_NUM '(' INTEGRAL ')'`, right?
So technically here `value.Type` should always be `IntValue`.

Reply:
Exactly